### PR TITLE
librewolf: use mkFirefoxModule

### PR DESCRIPTION
**Pull request by chayleaf**
### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

@onny

### CODE REVIEW

**Comment by tomodachi94** [comment]
Bookmarks don't appear to work, so they should probably be turned off somehow.

```nix
  programs.librewolf.profiles.default.bookmarks = [{
    name = "bookmarklets";
    toolbar = true;
    bookmarks = [
      {
        name = "example.com";
        url = "https://example.com";
      }
    ];
  }];
```

I vaguely recall someone mentioning this somewhere.

**Comment by brckd** [comment]
Yea I had shared the same experience with bookmarks in https://github.com/nix-community/home-manager/pull/5128#issue-2182871606, but removed it in a later edit of the message.

The option could be marked as `internal` to hide it from generated documentation and to discourage users from using it. Alternatively the `readOnly` option could be used with an empty default.

**Comment by brckd** [comment]
Looks great otherwise!

**Comment by chayleaf** [comment]
why does it not work though? it doesn't look like librewolf patches out any bookmark-related code

**Review by donovanglover — APPROVED** [review]
Nice to see this finally be a feature. From my initial testing everything works as expected.

For anyone else trying this out, the default search engine won't change in v128 unless you patch https://github.com/nix-community/home-manager/pull/5685 in your fork as well.

**Review by brckd — APPROVED** [review]

**Comment by SpiderUnderUrBed** [comment]
How would this implement nativeMessagingHosts and how could i use this for that when this gets merged?

**Comment by brckd** [comment]
It's implemented in the underlying `mkFirefoxModule`.
 https://github.com/nix-community/home-manager/blob/master/modules/programs/firefox/mkFirefoxModule.nix#L39-L49

https://github.com/nix-community/home-manager/blob/master/modules/programs/firefox/mkFirefoxModule.nix#L830-L834

Custom native messaging hosts can be added using `programs.librewolf.nativeMessagingHosts`.

File: modules/programs/librewolf.nix [hunk 1]
```diff
@@ -14,20 +14,33 @@ let
...
-      defaultText = literalExpression "pkgs.librewolf";
-      description = "The LibreWolf package to use.";
-    };
+  imports = [
+    (mkFirefoxModule {
+      inherit modulePath;
+      name = "LibreWolf";
+      wrappedPackageName = "librewolf";
+      unwrappedPackageName = "librewolf-unwrapped";
+      visible = true;
```
[the remark this answers is not in the document]
Comment (brckd) [reply modules/programs/librewolf.nix:30]: I don't know about the conventions, but that's how the chromite module does it. It also hides generic options for its derivatives.

**Review by MithicSpirit — APPROVED** [review]
Been using this and #5685 for a couple of days now, and it's been working very well.

**Comment by Anomalocaridid** [comment]
Is it normal for the tests to take this long? It seems to me like they have said "Waiting for status to be reported" since the beginning of last month.

**Comment by brckd** [comment]
Hey @chayleaf, in case you aren't busy, could you take a look at my latest review? This way we would give the tests a rerun.

**Comment by brckd** [comment]
These tests sure take long. Perhaps they haven't started yet because of GitHub limits?

**Comment by SpiderUnderUrBed** [comment]
I don't think these tests are going to finish anytime soon, is there something wrong with them? Or is there any possibility of running them locally if that will help.

**Comment by donovanglover** [comment]
The tests won't run until the [workflow is approved](https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-workflow-runs/approving-workflow-runs-from-public-forks).

**Comment by TilCreator** [comment]
I tried this PR and got an error:
```
home-manager-path> error: collision between `/nix/store/r4af2fhjq0m7sz20fx6apbdj61wzd7yg-librewolf-131.0.2-1/lib/librewolf/distribution/policies.json' and `/nix/store/9fbmav58r0mxlqwrxglz7lwzm3733c6g-librewolf-131.0.2-1/lib/librewolf/distribution/policies.json'
```
(the only difference of the two files is that one has `{"policies": {"DefaultDownloadDirectory": "/tmp"}}` set and the other has not)

I limited my librewolf config to
```
programs.librewolf = {
  enable = true;
  policies.DefaultDownloadDirectory = "/tmp";
};
```
which still reproduced the error, this noticeably does not fail when replacing `librewolf` with `firefox` in this minimal example.

A workaround is:
```
programs.librewolf = {
  enable = true;
  package = pkgs.librewolf.override (old: {
    extraPolicies = (old.extraPolicies or { }) // {DefaultDownloadDirectory = "/tmp";};
  });
};
```

all using `github:NixOS/nixpkgs/c4c95cac8161486a52f901a27ccdb5f8e80380b1`

**Comment by SpiderUnderUrBed** [comment]
Is it possible to declare different profiles using mkFirefoxModule? (by profiles i am referring to about:profiles)

**Comment by donovanglover** [comment]
If you mean multiple profiles in librewolf you can replace `firefox` with `librewolf` in the [docs](https://nix-community.github.io/home-manager/options.xhtml#opt-programs.firefox.profiles) and it will work

**Comment by Anomalocaridid** [comment]
@sid115 I think this is addressed by #5685

**Comment by MithicSpirit** [comment]
I have a similar issue to https://github.com/nix-community/home-manager/pull/5684#issuecomment-2423372447, but over the presence of `"ExtensionSettings":{}` in the policies file instead. Reverting #5840 fixed it, but I suspect that the fundamental issue is that it's even trying to include two librewolfs in the path in the first place.

**Comment by Anomalocaridid** [comment]
Could someone please approve the test workflow again? This has been sitting for a few weeks waiting for the tests to be re-run.

File: modules/programs/librewolf.nix [hunk 2]
```diff
@@ -53,7 +68,7 @@ in {
 
     home.packages = [ cfg.package ];
```
Comment (donovanglover) [inline modules/programs/librewolf.nix:69]:
```suggestion
```
Had to remove this to fix a collision error

**Comment by rycee** [comment]
Thanks! Merged to master now 🙂

**Comment by TilCreator** [comment]
Is it expected to get this feature (backported) in release-24.11? Would be really awesome to have this in the new release

**Comment by johnhamelink** [comment]
So after working with the Librewolf configuration for a while, it seems to be working well for me with the one exception of `nativeMessagingHosts` (I'm using `tridactyl-native` and `ff2mpv-rust`). As @TilCreator mentions above, after moving the config into a package override as described above, Librewolf does then interact with these packages.

```nix
  programs.librewolf = {
    enable = true;

    package = pkgs.librewolf.override {
      nativeMessagingHosts = with pkgs; [
        # Convert Firefox URLs to MPV runs
        ff2mpv-rust

        # Native messenger for tridactyl to enable all the features.
        # You can confirm it's found by FF by running `:native`.
        tridactyl-native
      ];
    };


    policies = (import ./policies.nix { });

    profiles.johnhamelink = (
      import ./profiles/johnhamelink.nix {
        inherit (pkgs) nixos-icons;
        inherit (pkgs.nur.repos.rycee) firefox-addons;
        inherit (lib) licenses platforms;
        inherit buildFirefoxXpiAddon;
      }
    );
  };
```

**Pull request merged**